libs/auth/src/openssl.ts

```diff
@@ -186,68 +211,254 @@ export async function verifySignature({
+ * the private key of the public key to sign was unavailable at the time of CSR creation, and a
+ * throwaway private key was used.
+ */
+export async function createCertFromCertSigningRequest({
+  certPublicKeyOverride,
+  certSigningRequest,
+  certType = 'standard_cert',
+  expiryInDays,
+  signingCertAuthorityCertPath,
+  signingPrivateKey,
+}: {
+  certPublicKeyOverride?: Buffer;
+  certSigningRequest: Buffer;
+  certType?: CertType;
+  expiryInDays: number;
+  signingCertAuthorityCertPath: string;
+  signingPrivateKey: FileKey | TpmKey;
+}): Promise<Buffer> {
+  const certAuthorityKeyParams: OpensslParam[] = (() => {
+    switch (signingPrivateKey.source) {
+      case 'file': {
+        return ['-CAkey', signingPrivateKey.path];
+      }
+      case 'tpm': {
+        return [
+          '-CAkey',
+          TPM_KEY_ID,
+          '-CAkeyform',
+          'engine',
+          '-engine',
+          OPENSSL_TPM_ENGINE_NAME,
+          '-passin',
+          `pass:${TPM_KEY_PASSWORD}`,
+        ];
+      }
+      /* istanbul ignore next: Compile-time check for completeness */
+      default: {
+        throwIllegalValue(signingPrivateKey, 'source');
+      }
+    }
+  })();
   const cert = await openssl([
     'x509',
     '-req',
     '-CA',
-    signingCertAuthorityCert,
-    '-CAkey',
-    signingPrivateKey,
-    '-passin',
-    `pass:${signingPrivateKeyPassword}`,
+    signingCertAuthorityCertPath,
+    ...certAuthorityKeyParams,
     '-CAcreateserial',
     '-CAserial',
     '/tmp/serial.txt',
     '-in',
     certSigningRequest,
-    '-force_pubkey',
-    publicKeyToSign,
     '-days',
     `${expiryInDays}`,
-    ...(certType === 'certAuthorityCert'
+    ...(certPublicKeyOverride ? ['-force_pubkey', certPublicKeyOverride] : []),
+    ...(certType === 'cert_authority_cert'
       ? ['-extensions', 'v3_ca', '-extfile', OPENSSL_CONFIG_FILE_PATH]
       : []),
   ]);
   return cert;
 }
+
+/**
+ * The input to createCert. All file keys and inline keys should be in PEM format.
+ */
+export interface CreateCertInput {
+  /**
+   * A private key should be provided when possible, but if the private key of the public key to
+   * sign isn't available (as is the case for a key pair generated on a Java Card), the public key is
+   * also acceptable. We set the cert key using -force_pubkey in this case.
+   */
+  certKeyInput:
+    | { type: 'private'; key: FileKey }
+    | { type: 'public'; key: InlineKey };
+  certSubject: string;
+  certType?: CertType;
+  expiryInDays: number;
+  signingCertAuthorityCertPath: string;
+  signingPrivateKey: FileKey | TpmKey;
+}
+
+const CreateCertInputSchema: z.ZodSchema<CreateCertInput> = z.object({
+  certKeyInput: z.union([
+    z.object({ type: z.literal('private'), key: FileKeySchema }),
+    z.object({ type: z.literal('public'), key: InlineKeySchema }),
+  ]),
+  certSubject: z.string(),
+  certType: z
+    .union([z.literal('cert_authority_cert'), z.literal('standard_cert')])
```

Comment:
Is there a way to derive this from `CertType`?

Reply:
As far as I can tell, there's not, looking at other Zod schemas for unions of string literals in the codebase:

https://github.com/votingworks/vxsuite/blob/8db46dceba2bb1c96c7e248e9896b5cb69c16ab8/libs/types/src/election.ts#L155-L159

That said, because we type the schema as:

```ts
const CreateCertInputSchema: z.ZodSchema<CreateCertInput> = z.object({
```

TypeScript will complain if the Zod schema gets out of sync in any way with the type it's for. So we don't have to worry about accidental regressions!